### ISSUE TITLE
A question about LayerNorm

### ISSUE THREAD

**Issue by mingyr** [deepmind/sonnet#68]
I noticed sonnet has provided a module called LayerNorm, which is based on the article https://arxiv.org/abs/1607.06450.
However, when I look at the corresponding implementation of TensowFlow, it provides a class called LayerNormBasicLSTMCell (TensorFlow does provides a standalone version of layer normalization tf.contrib.layers.layer_norm).
According to my understanding, and when I check the code in https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/contrib/rnn/python/ops/rnn_cell.py:

```

    i, j, f, o = array_ops.split(value=concat, num_or_size_splits=4, axis=1)
    if self._layer_norm:
      i = self._norm(i, "input")
      j = self._norm(j, "transform")
      f = self._norm(f, "forget")
      o = self._norm(o, "output")


```

What I mean here is the layer normalization is done INSIDE the cell, before activation. However, when sonnet provides the solo LayerNorm module in sonnet, in which way it is intended to be used? Like the following way:

```
import numpy as np
import tensorflow as tf
import sonnet as snt

class Model(snt.AbstractModule):
    def __init__(self, name = "model"):
        super(Model, self).__init__(name = name)

        hidden_size_array = [8, 8, 8]
        with self._enter_variable_scope():
            self._layer1 = snt.LSTM(hidden_size_array[0], name = 'layer1_lstm')
            self._norm1 = snt.BatchApply(snt.LayerNorm(), name = 'layer1_norm')

            self._layer2 = snt.LSTM(hidden_size_array[1], name = 'layer2_lstm')
            self._norm2 = snt.BatchApply(snt.LayerNorm(), name = 'layer2_norm')

            self._layer3 = snt.LSTM(hidden_size_array[2], name = 'layer3_lstm')
            self._norm3 = snt.BatchApply(snt.LayerNorm(), name = 'layer3_norm')


    def _build(self, inputs):

        batch_size = inputs.get_shape()[0]

        initial_state = self._layer1.initial_state(batch_size)
        output_sequence, final_state = \
            tf.nn.dynamic_rnn(self._layer1, inputs, initial_state = initial_state, time_major = False)

        output_sequence = self._norm1(output_sequence)

        initial_state = self._layer2.initial_state(batch_size)
        output_sequence, final_state = \
            tf.nn.dynamic_rnn(self._layer2, output_sequence, initial_state = initial_state, time_major = False)

        output_sequence = self._norm2(output_sequence)

        initial_state = self._layer3.initial_state(batch_size)
        output_sequence, final_state = \
            tf.nn.dynamic_rnn(self._layer3, output_sequence, initial_state = initial_state, time_major = False)

        output_sequence = self._norm3(output_sequence)

        return output_sequence

def test():

    t = tf.constant(np.ones((4, 6, 8)), tf.float32)
    model = Model()
    outputs = model(t)

    with tf.Session() as sess:
        sess.run([tf.global_variables_initializer()])

        v = sess.run(outputs)
        print(v)

if __name__ == "__main__":
    test()

```

Nevertheless in this way it can work, however, seems layer normalization is done outside the activation of LSTM cell. 

So I just wonder the dear engineers at DeepMind could give a doc a little detailed with regarding to the usage of LayerNorm module or not? (Or even probably my understanding of layer normalization is wrong). But I would like to hear the opinion of engineers there, so by clarification we can have sonnet easier to use.

Thanks a lot.

**Comment by malcolmreynolds** [comment]
Hi @mingyr - we provided LayerNorm as a separate module so that people can try using it in various places, without being tied to a particular RNN architecture. There is a constructor kwarg `use_layer_norm` for the LSTM which inserts a LayerNorm module inside the LSTM, before the activations, and so if you want to get the same behaviour as LayerNormBasicLSTMCell you can use that.

**Comment by mingyr** [comment]
@malcolmreynolds : Got it, great thanks for the detailed explanation.
